### PR TITLE
`check` / `match` / `react`

### PR DESCRIPTION
**Pull request by prismofeverything**
We have added three fundamental operations to the type system, two of which require a new registry as well (check/react). 

* `check` - Given a schema and any value, returns `True` if the value matches the schema, `False` otherwise. This also requires a registry for each built in type to make this distinction at the language level. 
* `match` - Given a schema, a state tree, and a pattern (the "redex"), find paths in the tree that match the pattern. There are a few modes for this depending on how many results you are interested in (immediate/first/random/all)
* `react` - This is an implementation of Milner's bigraphical reactions. Given a schema, a state tree, and a "reaction" (which contains either a redex/reactum pair or a reference to a function in the `react` registry that generates the redex/reactum), return the tree with matching redexes substituted with the corresponding reactum, according to the `match` mode (immediate/first/random/all). This defaults to `random` according to Milner's original formalism, though in practice we will probably use `first` more often (?)

A number of reactions are provided (add/remove/replace), with a notable exception (divide). A future PR will address reactions that require a change of state as well as structure during the reaction (which goes beyond bigraphical reactions which deal only in structure), as required by divide and many other useful reactions yet to be conceived. 

All feedback welcome!

### CODE REVIEW

**Comment by prismofeverything** [comment]
One more thing: due to the egregious imports between `base_types.py` and `type_system.py` I have pulled all the functions from `base_types.py` into `type_system.py`.... we are planning a future reorganization for the specification of types which will address this.

File: bigraph_schema/type_system.py [hunk 1]
```diff
@@ -572,6 +790,128 @@ def project_edge(self, schema, instance, edge_path, states, ports_key=None):
             states)
 
 
+    def infer_wires(self, ports, state, wires, top_schema=None, path=None):
```
Comment (eagmon) [inline bigraph_schema/type_system.py:793]: interesting. Does this map from ports to existing stores? does it make new stores at the same level? Does it look down into inners for matching stores?

File: bigraph_schema/type_system.py [hunk 2]
```diff
@@ -572,6 +790,128 @@ def project_edge(self, schema, instance, edge_path, states, ports_key=None):
...
+        '''
+        Given a schema fragment and an existing state with _type keys,
+        return the full schema required to describe that state,
+        and whatever state was hydrated (edges) during this process
+        '''
+
+        schema = schema or {}
+        # TODO: deal with this
+        if schema == '{}':
+            schema = {}
```
Comment (eagmon) [inline bigraph_schema/type_system.py:843]: yeah, that shouldn't happen.

File: bigraph_schema/type_system.py [hunk 3]
```diff
@@ -585,8 +925,563 @@ def query(self, schema, instance, redex):
...
+        return isinstance(instance, instance_type)
+
+    return is_instance_parameter
+
+
+check_number = instance_parameter(numbers.Number)
+check_boolean = instance_parameter(bool)
+check_int = instance_parameter(int)
+check_float = instance_parameter(float)
+check_string = instance_parameter(str)
```
Comment (eagmon) [inline bigraph_schema/type_system.py:942]: what's this over here?

**Review by eagmon — APPROVED** [review]
heck yes! These are critical new methods. I noticed some syntax such as the use of `@` and `?` in the new tests -- not sure if this is new syntax or if I've lost track. We need some way to document these all in one place, maybe in a Jupyter notebook.

**Pull request merged**